### PR TITLE
Searching and pagination for internal tables

### PR DESCRIPTION
**Pull request by aptkingston**
## Description
This PR builds upon @mike12345567's lab day project of implementing performing proper server side searching. Data providers can now be configured to search and filter data on the server, and optionally paginate their responses.

### Searching
When a data provider is configured to use an internal table, the `Filter` setting will open a drawer with advanced searching and filtering capabilities.

Here's an example with lots of expressions.
image

Different field types offer different matching operators, such as `equals`, `not equals`, `more than`, `less than`, `like`, `starts with`, `is empty` and `is not empty`. 
image

Values to match against can be either absolute values or they can be bindings. Because certain components such as datepickers cannot allow binding strings, there needs to be an explicit choice of a "Value" or a "Binding" to match against. Picking a value will show the correct component for the field type - e.g. a datepicker for a date field, or an options dropdown for an options field with the real options inside. Picking a "Binding" will show a bindable input, allowing users to open the binding drawer and enter a binding expression as usual.
image

If the data source is not an internal table (so a view, relationship or query) then a more basic version of search is offered, which will simply perform in-memory searching in the browser.

### Sorting
Server side sorting has also been implemented, allowing results to be sorted before being returned. This is especially useful when used in combination with pagination. Both sort column and sort order settings are available on data providers.
image

Data providers will automatically determine the best way to sort your data. If the sort column is of type number then it will use numeric sorting, otherwise it will use string sorting.

### Pagination
Pagination is controlled through the `Limit` and `Paginate` settings on a data provider.
image

The paginate checkbox is exactly what it sounds like, it controls whether or not search responses will be paginated. If this is checked, then the `Limit` setting is used as the page size. This is capped at 200, as this is a hard limitation in IBM's implementation of lucene. If pagination is disabled then the limit cap is raised to 1000. This is handled by fetching as many search pages as required until the limit is reached and then concatenating the results together, so this is transparent to the client. The cap of 1000 here is to avoid extremely long loading times or big performance hits, as even increasing this to 5k results increases load times in the order of ~5 seconds in my testing. Pages have to be fetched sequentially and synchronously which is the problem here. The default settings of a data provider are pagination enabled, limit 50.

If the data source is not an internal table then the pagination setting is ignored, but the results will still respect the `Limit` setting and be truncated appropriately.

When pagination is enabled, a new spectrum pagination control will be rendered at the bottom right of your data provider.
image

As lucene does not know how many pages of results there are, we cannot show the total number of pages. I've implemented a check of whether there is a following page by requesting 1 more row after every search and seeing if there is a result. Using this method we can at least inform the client of whether there is another page, so that page navigation buttons can be disabled and UX improved.

### Other misc changes
- I've added an animated spectrum loading spinner to data providers which provides a nicer loading experience than flashing empty states. This is especially useful when requesting large searches, such as for rendering a chart with 1000 points.
image.
- List autoscreens look exactly the same as before, although we can now change this if we desire. There will still be the new cap of 1000 in place, so we might want to change the default to be paginated instead. This is up for discussion, but won't cause any issues leaving it the way it is for now.
- I've removed the `Search` component and the corresponding endpoint, since its feature set is entirely included in the data provider implementation now.
- Added an optional `info` setting for components in the standard component manifest, which will display some text below the components settings. I've used this to explain that pagination only affects internal tables.
image

### Bug fixes
- Fix combobox disabled state not working
- Fix chart placeholder not displaying when inside the builder
- Fix spectrum background colours not being consistent
- Fix bug preventing progress circles from animating

### Future work
This opens up some very nice possibilities for improving the backend UI now, such adding proper search and pagination to the backend tables. We'll need to do this eventually, because there needs to be a total removal of any endpoints returning all rows in a table if we want to be totally scalable.

There are also some new opportunities now for making new components for things like filtering dashboards, which I'm not saying too much about because it's probably going to be my next lab day project. 🤫

### CODE REVIEW

**Comment by joebudi** [comment]
@aptkingston awesome work. I think we need to more explicit with the info content. When I first read it, I thought you were saying pagination was only possible with tables (component). Maybe we should say, 'The pagination feature is only available for data stored in internal tables.'

**Comment by aptkingston** [comment]
> @aptkingston awesome work. I think we need to more explicit with the info content. When I first read it, I thought you were saying pagination was only possible with tables (component). Maybe we should say, 'The pagination feature is only available for data stored in internal tables.'

Yep makes sense.

File: packages/server/src/db/views/staticViews.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ const SCREEN_PREFIX = DocumentTypes.SCREEN + SEPARATOR
  * @returns {Promise<void>} The view now exists, please note that the next view of this query will actually build it,
  * so it may be slow.
  */
-exports.createLinkView = async appId => {
+exports.createLinkView = async (appId) => {
```
Comment (mike12345567) [inline packages/server/src/db/views/staticViews.js:28]: think you might need to run a yarn format or something after pulling in next (after updating the eslint/prettier options).

**Review by mike12345567 — APPROVED** [review]
LGTM! Looks great, so amazing to have this functionality in the data provider, its gonna really open some doors. Really like the implementation of this generally too, I think its turned out pretty clean in the front and backend, for something that could have turned into a nightmare I think its turned out very clean! I've been talking about how critical pagination and searching was going to be to the performance of the system at scale, so happy to see this in as we should be able to scale to very large tables now - just need to fix the backend UI.

Also I think the backend UI rework with the ability to search - this would be amazing as you could use it for data administration then, a task that can sometimes get a little difficult in production systems - this would allow making changes to the production data easily in the development app, then deploying this rather than actually having to make changes live.

File: packages/server/src/db/views/staticViews.js (hunk 1 above)
Comment (aptkingston) [reply packages/server/src/db/views/staticViews.js]: I did indeed, thank you.

**Review by shogunpurple — APPROVED** [review]
Awesome work - this lays an excellent foundation for the next level of our internal DB/backend UI. 

The ability to use a query builder and saved searches/more powerful views in the backend UI paired with more powerful handlebars autocomplete and scripting capabilities will be absolutely killer 🚀

File: packages/server/src/api/controllers/search/utils.js [hunk 2]
```diff
@@ -73,50 +88,102 @@ class QueryBuilder {
...
     }
-    return buildSearchUrl({
-      appId: this.appId,
-      query: output,
-      bookmark: this.bookmark,
-      limit: this.limit,
-    })
+
+    // Build the full search URL
+    let url = `${env.COUCH_DB_URL}/${this.appId}/_design/database/_search`
```
[the remark this answers is not in the document]
Comment (aptkingston) [reply packages/server/src/api/controllers/search/utils.js:161]: Nice find, just saw that hidden in the IBM docs. I'll update that, would be much cleaner.

**Pull request merged**